packages/cloud_firestore/ios/Classes/CloudFirestorePlugin.m

```diff
@@ -112,7 +114,93 @@ - (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result
+      NSNumber *transactionTimeout = call.arguments[@"transactionTimeout"];
+
+      self.transactions[transactionId] = transaction;
+
+      dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
+
+      [self.channel invokeMethod:@"DoTransaction"
+                       arguments:call.arguments
+                          result:^(id doTransactionResult) {
+                            self.transactionResults[transactionId] = doTransactionResult;
+                            dispatch_semaphore_signal(semaphore);
+                          }];
+
+      dispatch_semaphore_wait(
+          semaphore, dispatch_time(DISPATCH_TIME_NOW, [transactionTimeout integerValue] * 1000000));
+
+      return self.transactionResults[transactionId];
+    }
+        completion:^(id transactionResult, NSError *error) {
+          if (error != nil) {
+            NSLog(@"This is the transaction error: %@", error.localizedDescription);
+            result([FlutterError errorWithCode:[NSString stringWithFormat:@"%ld", error.code]
+                                       message:error.localizedDescription
+                                       details:nil]);
+          }
+          result(transactionResult);
+        }];
+  } else if ([@"Transaction#get" isEqualToString:call.method]) {
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+      NSNumber *transactionId = call.arguments[@"transactionId"];
+      NSString *path = call.arguments[@"path"];
+      FIRDocumentReference *documentReference = [[FIRFirestore firestore] documentWithPath:path];
+      FIRTransaction *transaction = self.transactions[transactionId];
+      NSError *pError = [[NSError alloc] init];
+
+      FIRDocumentSnapshot *snapshot = [transaction getDocument:documentReference error:&pError];
+
+      if (pError != nil) {
+        result([FlutterError errorWithCode:[NSString stringWithFormat:@"%tu", [pError code]]
+                                   message:[pError localizedDescription]
+                                   details:nil]);
+      } else if (snapshot != nil) {
+        result(@{
+          @"path" : snapshot.reference.path,
+          @"data" : snapshot.exists ? snapshot.data : [NSNull null]
+        });
+      } else {
+        result([FlutterError errorWithCode:@"DOCUMENT_NOT_FOUND"
+                                   message:@"Document not found."
+                                   details:nil]);
+      }
+    });
+  } else if ([@"Transaction#update" isEqualToString:call.method]) {
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+      NSNumber *transactionId = call.arguments[@"transactionId"];
+      NSString *path = call.arguments[@"path"];
+      FIRDocumentReference *documentReference = [[FIRFirestore firestore] documentWithPath:path];
+      FIRTransaction *transaction = self.transactions[transactionId];
+
+      [transaction updateData:call.arguments[@"data"] forDocument:documentReference];
```

Comment:
It seems a bit odd to me that this can throw exceptions that we handle on Android but not iOS, but maybe it's fine.

Reply:
It does seem strange to me as well. [Documentation](https://github.com/firebase/firebase-ios-sdk/blob/7386f00178be4fe1f8d8f749049d25cbf0533e6a/Firestore/Source/Public/FIRTransaction.h#L69) says that if the document does not exist the transaction will fail.  So no exception to catch.